### PR TITLE
KFSPTS-20442 Add create-acct-doc duplicate file checks

### PR DESCRIPTION
**Pull request by cah292**
There are PRs for this in both cu-kfs and nonprod-sql. Please make sure both PRs are good to go before merging.

Also, I'm waiting for functional feedback on some of the new messages introduced by these changes. Please don't merge until after I hear back about any needed wording changes.

This PR allows the Create Accounting Documents Job to prevent the processing of files that were already handled in a previous run; such functionality is controllable by a new parameter. I believe I've implemented the requirements noted in the user story, but please let me know if I missed anything. (Adding the new table to the Purge Job will be handled in a follow-up user story, as noted in the ticket.)

I also fixed a pre-existing bug that was preventing certain error information from being displayed properly in the reports. (I needed that functionality for the duplicate file error reporting.) This partially fixes the KFSPTS-30149 issue, but that ticket still needs more follow-up work.

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:336]: Should this boolean expression just be reportItem.isDuplicateFile()?  We only send email in production, and we will always prevent duplicate in prod.  Also, if duplicates are are found, XML won't fully be loaded, correct?

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:362]: I think we want DUPLICATE_FILE_REPORT_EMAIL_ADDRESS to have a list of addresses.  We want the developers to know because there may be something technically going awry, and the functionals (FIS / Cathy) would also want to know

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:380]: I am thinking about the situation where the XML file is fine, and fully loaded the KFS documents, but the report email address is empty or not valid.  I think we need to alert the developers that we received a file without a report email address.

File: src/main/resources/edu/cornell/kfs/fp/cu-ojb-fp.xml
Comment (jhulslander) [inline src/main/resources/edu/cornell/kfs/fp/cu-ojb-fp.xml:1331]: sorry I didn't mention it before, but it may be nice to include sequence ID field as the primary key, and make FIL_NM a unique constraint.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:209]: similar to the length check on the overview, I think we want a length check on the email and file name

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:198]: I find this convention a bit confusing, and perhaps error prone with potential future updates.  If an XML was previously loaded, we don't want to update the values for existingEntry.  This naming convention makes it sound like that is possible.  Can this be refactored to help with that?

**Review by jhulslander — CHANGES_REQUESTED** [review]
Changes are looking good, but I found a few items I think need to be looked at

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:57]: I'll add max length checks on the other fields accordingly.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:96]: There's some logging down in the "catch" block, but I'll see if I can move most of that logging over into this block instead.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:137]: Adding the fileEntry to the log statement will cause its toString() content to be added in the appropriate placeholder. I purposely overrode the BO's toString() method so that it would print the BO's properties, in case we needed that information in the logs to assist with troubleshooting.

If you think we shouldn't add that info to the logs, please let me know.

Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:137]: That makes sense, and I agree, thank you

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:336]: I think we need at least the `isDuplicateFile()` and `shouldExcludePreviouslyProcessedFiles()` checks here. If the parameter is configured to allow duplicate processing, the report item will be marked as a duplicate but KFS will process it anyway. I don't think we want the extra duplicate-file recipients to be notified if KFS has been configured to allow such duplication. If you think otherwise, please let me know.

I believe I could at least simplify this to remove the `!reportItem.isXmlSuccessfullyLoaded()` check.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:380]: The enclosing `getReportEmailFromFileIfPossible()` method is only meant to be used in cases where the file was rejected due to the duplicate-file check. We already have existing code in `CreateAccountingDocumentValidationServiceImpl` that will validate the correctness of the report email, if the file was not rejected by the duplication check. An email validation error at that phase will cause a header-level validation failure in the file, and will cause the error report to be sent to the email specified by `CREATE_ACCT_DOC_REPORT_EMAIL_ADDRESS`.

If you think we need to adjust the existing code so that such email error reports are also sent to a developer EGA, please let me know.

File: src/main/resources/edu/cornell/kfs/fp/cu-ojb-fp.xml
Comment (cah292) [reply src/main/resources/edu/cornell/kfs/fp/cu-ojb-fp.xml:1331]: I'll add the suggested columns and constraints.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:209]: I'll add the extra length checks accordingly.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:198]: I can modify this to only support creating entries and not updating them. I wasn't sure if we needed to allow for updating existing entries when duplication is permitted, but if you're fine with the file entries only reflecting the first processed file of a given name (as opposed to the latest file), then I'm fine with making this a create-only method.

**Comment by cah292** [comment]
I think I've updated the areas that needed further changes. This is ready for review again.

I have a question out to the functionals as to whether any extra emails need to be included as duplicate-file-error recipients. Once I get feedback on that and make any necessary follow-up changes, this can be merged once it has been reviewed and approved.

**Comment by jhulslander** [comment]
retest this please

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java]: I don't think we want to adjust previously saved records, they should be a historic record of a processed file

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:391]: that should be fine, thanks

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:215]: I think the file size clean should be included in getCleanedFileName() so the search will find entries of file names that had to be trimmed for length

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:211]: Is this needed?  I thought BusinessObjectService / OJB would set this if not set.

**Review by jhulslander — COMMENTED** [review]
Changes are looking good, i found a few more things to consider

**Comment by jhulslander** [comment]
retest this please

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:215]: I'll refactor the cleaning and truncating of the filename accordingly.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:211]: I'll see if I can set this up through OJB metadata instead of the SequenceAccessorService.

Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:211]: I don't think you need to set the fileId value, won't OJB set it atomically from the sequence?

Comment (cah292) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java:211]: Yes, I meant to say that I'm changing this to use just the OJB XML for this setup, and that I'll remove the manual ID-populating code (at least on the non-unit-test side). Sorry for the confusion.

**Comment by cah292** [comment]
I've committed some more changes to address the latest feedback. (A small change has also been committed to the SQL PR.) This is ready for review again.

Also, I've received the needed functional feedback on my questions, so this is now ready to be merged once both PRs have been approved.

**Review by jhulslander — APPROVED** [review]
code changes look good, as do SQL changes, unit tests pass, I shall merge

**Pull request merged**
